modules/reindex/src/main/java/org/elasticsearch/index/reindex/remote/RemoteRequestBuilders.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.reindex.remote;
+
+import org.apache.http.HttpEntity;
+import org.apache.http.entity.ByteArrayEntity;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.StringEntity;
+import org.elasticsearch.Version;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.search.sort.FieldSortBuilder;
+import org.elasticsearch.search.sort.SortBuilder;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.util.Collections.singletonMap;
+
+final class RemoteRequestBuilders {
+    private RemoteRequestBuilders() {}
+
+    static class InitialSearch {
+        private InitialSearch() {}
+
+        static String path(SearchRequest searchRequest) {
+            // It is nasty to build paths with StringBuilder but we'll be careful....
+            StringBuilder path = new StringBuilder("/");
+            addIndexesOrTypes(path, "Index", searchRequest.indices());
+            addIndexesOrTypes(path, "Type", searchRequest.types());
+            path.append("_search");
+            return path.toString();
+        }
+
+        static Map<String, String> params(SearchRequest searchRequest, Version remoteVersion) {
+            Map<String, String> params = new HashMap<>();
+            if (searchRequest.scroll() != null) {
+                params.put("scroll", searchRequest.scroll().keepAlive().toString());
+            }
+            params.put("size", Integer.toString(searchRequest.source().size()));
+            if (searchRequest.source().version() == null || searchRequest.source().version() == true) {
+                // false is the only value that makes it false. Null defaults to true....
+                params.put("version", null);
```

Comment:
Would it be more efficient to skip fetching it if `version_type=external` isn't used?